frontend_tests/zjsunit/namespace.js

```diff
@@ -28,23 +30,12 @@ exports.start = () => {
     Module._load = load;
 };
 
-exports.mock_module = (short_fn, obj) => {
-    if (obj === undefined) {
-        obj = {};
-    }
-
-    if (typeof obj !== "object") {
-        throw new TypeError("We expect you to stub with an object.");
-    }
-
-    if (short_fn.startsWith("/") || short_fn.includes(".")) {
-        throw new Error(`
-            There is no need for a path like ${short_fn}.
-            We just assume the file is under static/js.
-        `);
-    }
-
-    const filename = require.resolve(`../../static/js/${short_fn}`);
+exports.mock_cjs = (request, obj) => {
```

Comment:
These probably deserve good comments/docstrings explaining how to decide which to use for contributors who don't understand the difference between CJS modules and ESM modules (which is probably most of them).

Reply:
Also, I wonder if we're likely to end up in a world where 90% of modules that exist are ESM, and we want to have one called `mock_module` and one called `mock_cjs_module`, if we're only using the `cjs` case for a handful of third-party modules by a month from now.

Reply:
Someone who needs to learn about CommonJS vs. ES6 modules will be even more confused if our names don’t reflect real concepts that they can go look up, or they wrongly suggest a subset relationship rather than a dichotomy, or they fail to suggest that an alternative exists if you only happen to see one. Brevity is important, but correctness is more important.

I could be wrong, but I’m guessing a library like jQuery is unlikely to ever expose an ES module, because then all of its CJS consumers would have to migrate from `$ = require("jquery")` to `{$} = require("jquery")`. So I don’t think there’s a near future where dealing with CJS modules is a rare exception.

I’ve added a comment introducing the difference between the formats. It’s not intended to be comprehensive (that could fill a book and some aspects are still evolving), but should be enough as a starting point for figuring out what to look up or to ask for help about.

Reply:
I am gonna state my argument mostly for the benefit of @timabbott.

We don't need a dichotomy here when a trichotomy can provide more clarity.

First some stats:
* there are 344 calls to mock_esm with modules under "../../static/js"
* there is 1 call to mock_esm that is outside of "../../static/js" (Sortable)
* there are 0 calls to mock_cjs with modules under "../../static/js"
* there are 62 calls to mock_cjs with modules outside of "../../static/js" (mostly jQuery, plus one each of clipboard and @uppy/core)

My concern with the 344 repetitions of "../../static/js" isn't the extra typing; it's the noise for the person reading the code.  I understand the explicitness concerns, but you could DRY all of these up with `mock_zulip(shorter_path)`.  I understand there are tradeoffs here, but sometimes it's just nice not to repeat the same thing 344 times and add a little wrapper.  I understand that the current ugliness is exaggerated by our current directory structure, but if we go to a new directory structure with a one-line wrapper in place, the future diff will be one line instead of 344+.

And then for external functions, we could have `mock_external({fn, module_type}` and require `module_type` to be "esm" or "cjs".  Or, if you prefer, `mock_external_cjs` and `mock_external_esm`.

I don't think my later proposal for much-added brevity would cause the correctness concerns that @andersk brought up.  If you're gonna mock a Zulip file, it's gonna be already converted.  And if you mock an external file, you will have to specify the module_type.

Reply:
My argument is also one of clarity and readability. To someone who hasn’t been working with them every day, the Zulip Node tests are full of unfamiliar concepts specific to the Zulip Node tests. A redundant trichotomy should not need to be one of them, nor should a custom system for abbreviating paths. Where you look at `../../static/js` and see noise, I look at it and see a beacon that someone can point at and go, *oh hey, that’s a module path—I know what that means*. And then they won’t have to spend extra effort translating that knowledge across differently spelled APIs that do the same thing.

I find it hard to be swayed by your complaint that I’m sweeping 344 modules (with plenty of opportunity for review), when that’s exactly what you did in #17514 (with no review, over my objections in #17439).

I split out a tiny commit to introduce `mock_cjs` after `mock_esm`. (I chose this ordering intentionally so that merging a prefix of this PR doesn’t leave us with a dichotomy that doesn’t make sense.)

Reply:
I think after thinking about this a bit more, I'm basically happy with the `mock_esm / mock_cjs` dichotomy.  We may want to promote those to topics for the "writing node tests" documentation, but to some extent folks will find it easy enough to pattern-match.